pandas/core/groupby.py

```diff
@@ -3740,9 +3740,30 @@ def _reindex_output(self, result):
             return result
 
         levels_list = [ping.group_index for ping in groupings]
-        index = MultiIndex.from_product(levels_list, names=self.grouper.names)
-        d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+        index, _ = MultiIndex.from_product(
+            levels_list, names=self.grouper.names).sortlevel()
+
+        if self.as_index:
+            d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
+            return result.reindex(**d)
+
+        # GH 13204
+        # Expanding
+        result = result.set_index(self.grouper.result_index).reindex(index)
+        # Now, column groupers are contained in `result` as both:
+        # - columns (containing NaN's after reindexing; will be dropped)
+        # - index levels (properly expanded; will be reset as columns).
+
+        # Column groupers only
```

Comment:
ok this looks a lot better.

Just give an other pass at the comments, pretend you are a new contributor and you are like, what the heck is going on here.